CONTRIBUTING: remove suggestion about the order of arguments

AndersonTorres opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

fricklerhandwerk commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -196,7 +196,20 @@
+  8. `buildPackages`;
+  9. Darwin-related arguments (e.g. `darwin`, `IOKit`, `Metal`).
+
+  Each item is optional, in the sense it should not be included when not needed (e.g. if there is no need for `buildPackages`, then it should not be included).
```
```suggestion
  Each item is optional, in the sense it should not be included when not needed. For example, if there is no need for `buildPackages`, then it should not be included.
```

Avoid nesting, it makes parsing harder.

fricklerhandwerk commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -196,7 +196,20 @@
+
+  Each item is optional, in the sense it should not be included when not needed (e.g. if there is no need for `buildPackages`, then it should not be included).
+
+  Inside each item of the hierarchy above, the arguments should be in "ASCIIbetical" order, (e.g. `SDL` comes before `Xaw3d`, that comes before `alsa-lib`, that comes before `assemblyscript`).
```
```suggestion
  Inside each item of the hierarchy above, the arguments should be in "ASCIIbetical" order.
Example:`SDL`, `Xaw3d`, `alsa-lib`, `assemblyscript`.
```

Just make a concrete example, avoid filling words that don't convey information - that it easier to scan. 

Is there a particular reason for this oddly specific convention though?

AndersonTorres commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -196,7 +196,20 @@
+  5. all the remaining arguments (i.e. those that don't fit the other items);
+  6. options (e.g. `extraModules`);
+  7. Boolean flags (e.g. `enableSDL`);
+  8. `buildPackages`;
```
I will put this before the other items, to be sure, between 4 and 5.

AndersonTorres reviewed (commented):
Remember myself.

figsoda commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -196,7 +196,23 @@
+  08. Boolean flags (`enableSDL`);
+  09. testers (`nixosTests`, `testers`);
+  10. updaters (`gitUpdater`, `gitUnstableUpdater`);
+  11. Darwin-related arguments (`darwin`, `Metal`, `IOKit`).
```
I don't think darwin should be special cased here, this can be merged into `all the remaining arguments`

AndersonTorres replied:
This is more because of tradition. They always appeared last, sometimes with a comment like `# Darwin dependencies`.

I believe this is good, because they are dealt only when some specific Darwin quirk comes into the picture.

Keeping Darwin dependencies in a dedicated block helps when someone proposes a Darwin-specific change.

figsoda commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -196,7 +196,23 @@
+
+  01. `lib`;
+  02. `stdenv`;
+  03. builders (`buildPythonPackage`, `mkDerivation`, `buildGoModule`);
```
"builders" needs to be defined here, does `rustPlatform` count? what about `makeRustPlatform`? what about `python3` if you are just using `python3.pkgs.buildPythonApplication`?

AndersonTorres replied:
> "builders" needs to be defined here, does `rustPlatform` count? what about `makeRustPlatform`?

Now I am a bit confused. Is `rustPlatform` something similar to `stdenv`?
There are other `languagePlatforms` similar to it, so I can compare?

For now I am inclined to include them here too.

> what about `python3` if you are just using `python3.pkgs.buildPythonApplication`?

This is still `python3`. So, item 6.

I believe this is consistent because you are not seeking for a typical builder, but by the Python-ecosystem-provided one.

figsoda replied:
`rustPlatform` is essentially a wrapper for `cargo`, `rustc`, `cargo-auditable`, and `stdenv`, and exposes things like `buildRustPackage` (a builder, which the most commonly directly used), `fetchCargoTarball` (fod fetcher that takes a `src`), and `rustLibSrc` (derivation)

AndersonTorres replied:
Basically a stdenv with rusteroids.
Fine. Let's lump them.

figsoda commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -196,7 +196,23 @@
+- The order of elements in the argument set must follow the convention below:
+
+  01. `lib`;
+  02. `stdenv`;
```
I would argue that this should be merged into the builders, if not, other types of stdenv should be mentioned as well, e.g. `clangStdenv`, `stdenvNoCC`

AndersonTorres replied:
"`stdenv` and friends" would be a good name?

Or you are also suggesting to include/lump stdenvs and builders here?

figsoda replied:
> Or you are also suggesting to include/lump stdenvs and builders here?

I would prefer that, yes

figsoda commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -196,7 +196,23 @@
+  02. `stdenv`;
+  03. builders (`buildPythonPackage`, `mkDerivation`, `buildGoModule`);
+  04. fetchers (`fetchurl`, `fetchpatch`, `fetchFromGitHub`);
+  05. `buildPackages`;
```
other sets should also be mentioned here, e.g. `pkgsBuildBuild` and `pkgs`

AndersonTorres replied:
Nice.

figsoda commented on doc/contributing/coding-conventions.chapter.md:
```diff
@@ -196,7 +196,23 @@
+  06. all the remaining arguments (i.e. those that don't fit the other items);
+  07. options (`extraModules`, `disabledProtocols ? [ "tcp"  "udp" ]`);
+  08. Boolean flags (`enableSDL`);
+  09. testers (`nixosTests`, `testers`);
```
A common pattern in tests is

```nix
testers.testVersion {
  package = <attribute>;
}
```

Since a lot of builders still doesn't support the `finalAttrs` pattern, `<attribute>` would need to be added to the inputs, maybe attributes like this should be added to this section as well?

zimbatm reviewed (dismissed):
I think this is a good idea, but style enforcement should be done on the tooling level. Otherwise, it adds more weight on both the author and reviewers of the PRs, and it distracts them from the substance of the changes.

AndersonTorres commented:
You say by the ofborg CI?

zimbatm commented:
yes, that kind of tool could be invoked by ofborg, or GitHub Actions

AndersonTorres commented:
But there is another coding convention being enforced already by the document - namely, "lib first; order of appearance otherwise".

Therefore, any external tool like nixpkgs-hammering is bound to this convention.

So, this paragraph should be removed, I believe.

AndersonTorres committed:
CONTRIBUTING: remove suggestion about the order of arguments

Particularities about style should be enforced at the tooling level (linters,
code beautifiers etc.). Otherwise, it adds more weight on the PR reviewing,
distracting both authors and reviewers from the substance of the changes.

infinisil reviewed (changes requested):
I agree with @zimbatm, either let's do it in tooling or not at all, I wouldn't want to force this fairly arbitrary convention on every reviewer and contributor.

AndersonTorres commented:
Then the paragraph should be removed by the same reason. I will update this.
